File: ocaml/21-30/21.ml
(* Insert an element at a given position into a list. (easy)

Start counting list elements with 0. If the position is larger or equal to the length of the list, insert the element at the end. (The behavior is unspecified if the position is negative.)
*)

let rec insert_at v i = function
  | [] -> [v]
  | x::xs -> if i = 0 then v::x::xs else x::(insert_at v (i-1) xs)

let _ = assert ((insert_at "alfa" 1 ["a";"b";"c";"d"]) = ["a"; "alfa"; "b"; "c"; "d"])
let _ = assert ((insert_at "alfa" 3 ["a";"b";"c";"d"]) = ["a"; "b"; "c"; "alfa"; "d"])
let _ = assert ((insert_at "alfa" 4 ["a";"b";"c";"d"]) = ["a"; "b"; "c"; "d"; "alfa"])

File: prolog/21-30/21.pl
/*
Insert an element at a given position into a list.
    Example:
    ?- insert_at(alfa,[a,b,c,d],2,L).
    L = [a,alfa,b,c,d]
*/


ins(X, [H|T], 1, Pre, Ys) :- 
  reverse(Pre, Ps),
  append(Ps, [X, H | T], Ys).
ins(X, [H|T], N, Pre, Ys) :- 
  N > 1,
  N1 is N - 1,
  ins(X, T, N1, [H|Pre], Ys).

insert_at(X, Xs, 1, [X|Xs]).
insert_at(X, [H|T], N, Ys) :- 
  N > 1,
  ins(X, [H|T], N, [], Ys).

